coding/pybind11_style_guide.rst

```diff
@@ -0,0 +1,584 @@
+    instead as soon as C++14 support is available.
+
+.. _style-guide-pybind11-holder-type:
+
+The shared_ptr holder type SHOULD be used for all non-trivial classes
+---------------------------------------------------------------------
+
+By not specifying a holder type explicitly it becomes ``unique_ptr``, but it is hard to anticipate when wrapping a class whether any downstream code will later use it with ``shared_ptr``.  Moreover, C++ functions taking ``unique_ptr`` arguments can never be wrapped intuitively in Python (because Python has no output arguments or ownership transfer), so we do not need to worry about wrapped instances held by ``shared_ptr`` that must be converted to ``unique_ptr`` for a function call.
+
+The only classes that should be wrapped with ``unique_ptr`` are non polymorphic classes that are always passed by value or reference in C++ and are small enough that ``shared_ptr`` represents a significant overhead.
+
+Note that this does not mean that ``shared_ptr`` must be used in C++ code in preference to other options; the :ref:`C++ coding guidelines on when to use them <style-guide-cpp-5-24b>` still apply.
+
+.. _style-guide-pybind11-keyword-arguments:
+
+Keyword names MUST be provided for functions with multiple overloads or more than two arguments
+-----------------------------------------------------------------------------------------------
+
+Keyword arguments make Python code significantly more readable, especially when distinguishing between overloads or in long function signatures.
+
+Keyword arguments MAY be provided for non-overloaded functions with two or fewer arguments, and are strongly encouraged if the meaning or order of the arguments is not apparent from the function name.
+
+.. _style-guide-pybind11-keyword-argument-literals:
+
+Literals MUST be used for all named arguments
+----------------------------------------------
+
+The `_a` argument literal, from `pybind11::literals` MUST be used
+for all named arguments (e.g. ``mod.def("f", f, "arg1"_a, "arg2"_a);``).
+The ``py::arg()`` construct SHALL NOT be used.
+
+.. _style-guide-pybind11-enum-scoping:
+
+Enum scoping SHALL follow usage in C++
+--------------------------------------
+
+* Unscoped enums SHALL export their names into the class scope using ``.export_values``:
+
+.. code-block:: cpp
+
+    py::enum_<Class::State>(cls, "State")
+        .value("RED", &Class::State::RED)
+        .value("GREEN", &Class::State::GREEN)
+        .export_values();
+
+* Scoped enums (i.e. ``enum class`` in C++) SHALL NOT use ``.export_values``.
+
+.. _style-guide-pybind11-enums-as-integers:
+
+Enums used as integers SHALL be wrapped as integer attributes
+-------------------------------------------------------------
+
+Regular (non-class) enums are frequently used in C++ to define a set of related integer constants rather than an actual enumeration.  These enums should be wrapped as simple integer class attributes rather than pybind11 enums, e.g.::
```

Comment:
I think it's worth explicitly mentioning bit flags here, as I think that's by far the most common case of enums used as integers, such as, "Enums whose values are defined to be distinct bits (e.g. `0x01, 0x02, 0x04`) are almost certainly used only as integer constants."